c/src/atom.rs

```diff
@@ -709,14 +806,8 @@ pub fn ptr_into_bindings(bindings: *mut bindings_t) -> Bindings {
     unsafe {Box::from_raw(bindings)}.bindings
 }
 
-fn vec_atom_into_ptr(vec: Vec<Atom>) -> *mut vec_atom_t {
-    Box::into_raw(Box::new(vec_atom_t(vec)))
-}
-
 pub fn return_atoms(atoms: &Vec<Atom>, callback: c_atoms_callback_t, context: *mut c_void) {
-    let results: Vec<atom_ref_t> = atoms.iter()
-        .map(|atom| atom.into()).collect();
-    callback((&results).into(), context);
+    callback(&(&atoms[..]).into(), context);
```

Comment:
Except `atom_get_children` case `return_atoms` is used to return owned instance of the `Vec<Atom>` maybe it makes sense split this function on two: `return_owned_atoms`, `return_borrowed_atoms`? Then in turn we should introduce to versions of `c_atoms_callback_t` and modify `hyperonpy.cpp` to clone atoms when borrowed instance is returned and move them when owned instance is returned.

Reply:
I don't understand.  The implementation of `atom_get_children` provides a borrowed reference to the atoms, mirroring the Rust interface.  Then when `atom_get_children` is called in hyperonpy.cpp, the returned atoms are already cloned in the `copy_atoms` callback function.

Are you saying you would like to see an additional interface that consumes an expression atom, and returns an owned vector of its children?

Reply:
> Are you saying you would like to see an additional interface that consumes an expression atom, and returns an owned vector of its children?

Sorry I mentioned `atom_get_children` and it is probably confused you. I mean that all usages of `return_atom` (there are about 5 places as far as I remember) except `atom_get_children` actually could pass ownership of the `Vec<Atom>` to the `return...` function. Thus we could introduce another version of the `return_atoms` which gets ownership and use this version everywhere except `atom_get_children`.